flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/service/application/ScriptExecutor.java

```diff
@@ -0,0 +1,332 @@
+            for (int i = position; i < script.length(); i++) {
+                char lastChar = currentChar;
+                currentChar = script.charAt(i);
+
+                currentSqlBuilder.append(currentChar);
+                currentPaddingLineBuilder.append(" ");
+
+                switch (currentChar) {
+                    case '\'':
+                        if (state == State.SINGLE_QUOTE) {
+                            state = State.NORMAL;
+                        } else if (state == State.NORMAL) {
+                            state = State.SINGLE_QUOTE;
+                        }
+                        break;
+                    case '"':
+                        if (state == State.DOUBLE_QUOTE) {
+                            state = State.NORMAL;
+                        } else if (state == State.NORMAL) {
+                            state = State.DOUBLE_QUOTE;
+                        }
+                        break;
+                    case '`':
+                        if (state == State.BACK_QUOTE) {
+                            state = State.NORMAL;
+                        } else if (state == State.NORMAL) {
+                            state = State.BACK_QUOTE;
+                        }
+                        break;
+                    case '-':
+                        if (lastChar == '-' && state == State.NORMAL) {
+                            state = State.SINGLE_COMMENT;
+                        }
+                        break;
+                    case '\n':
+                        if (state == State.SINGLE_COMMENT) {
+                            state = State.NORMAL;
+                        }
+                        currentPaddingLineBuilder.setLength(0);
+                        currentPaddingSqlBuilder.append("\n");
+                        break;
+                    case '*':
+                        if (lastChar == '/' && state == State.NORMAL) {
+                            state = State.MULTI_LINE_COMMENT;
+                        }
+                        break;
+                    case '/':
+                        if (lastChar == '*' && state == State.MULTI_LINE_COMMENT) {
+                            state = State.NORMAL;
+                        }
+                        break;
+                    case ';':
+                        if (state == State.NORMAL) {
+                            i =
+                                    prefetch(
+                                            i + 1,
+                                            currentSqlBuilder,
+                                            currentPaddingSqlBuilder,
+                                            currentPaddingLineBuilder);
+                            statement = currentSqlBuilder.toString();
+                            try {
+                                position = i + 1;
+                                parse(previousPaddingSqlBuilder + statement);
```

Comment:
It is used to detect whether the statement is completed. Because Flink SQL supports `EXECUTE STATEMENT STATEMENT SET BEGIN INSERT INTO ...;`, semicolon is not the end of the statments(`END;` is the real end.). So we need to validate the statement is completed or not.